### ISSUE TITLE
Misleading documentation in bbox_iou function for bounding box input

### ISSUE THREAD

**Issue by visionNoob** [ultralytics/ultralytics#18539]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Other

### Bug

Hi

I noticed that the docstring in the `bbox_iou` function appears to contain an incorrect description of the box1 argument. Specifically, the docstring says:

https://github.com/ultralytics/ultralytics/blob/00aefd795c13bb6a5364b63798c383adc15248d4/ultralytics/utils/metrics.py#L79

However, this doesn't align with the actual implementation. The function supports a batch of bounding boxes for `box1`, and `box1` can have a shape of `(n, 4)`, where `n >= 1.`

For example, this call works as expected:
```python
box1 = torch.tensor([[0.5, 0.5, 1.0, 1.0], [0.3, 0.3, 0.8, 0.8]])  # Shape (2, 4)
box2 = torch.tensor([[0.4, 0.4, 0.9, 0.9]])  # Shape (1, 4)
iou = bbox_iou(box1, box2, xywh=False)
print(iou)  # Outputs IoU values for both bounding boxes in `box1`
```
Plus, 
this is confirmed in the `TaskAlignedAssigner` class in `TAL.py`, where the `_forward` method uses `gt_bboxes` (which corresponds to `box1` in `bbox_iou`) with shapes like `(bs, max_num_object, 4)`—clearly indicating that box1 can hold multiple bounding boxes:

https://github.com/ultralytics/ultralytics/blob/00aefd795c13bb6a5364b63798c383adc15248d4/ultralytics/utils/tal.py#L155

If I’ve misunderstood anything, please let me know. Thanks again for the amazing work you’re doing!

### Environment

Ultralytics 8.3.56 🚀 Python-3.10.12 torch-2.5.1+cu124 CUDA:0 (NVIDIA GeForce RTX 3060, 12288MiB)
Setup complete ✅ (12 CPUs, 31.2 GB RAM, 39.1/250.9 GB disk)

OS                  Linux-5.15.146.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.12
Install             pip
RAM                 31.24 GB
Disk                39.1/250.9 GB
CPU                 12th Gen Intel Core(TM) i5-12600
CPU count           12
GPU                 NVIDIA GeForce RTX 3060, 12288MiB
GPU count           1
CUDA                12.4

numpy               ✅ 1.26.4>=1.23.0
numpy               ✅ 1.26.4<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.8.3>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.2.0>=7.1.2
pyyaml              ✅ 5.4.1>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.5.1>=1.8.0
torch               ✅ 2.5.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.20.1>=0.9.0
tqdm                ✅ 4.66.2>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.13>=2.0.0
{'OS': 'Linux-5.15.146.1-microsoft-standard-WSL2-x86_64-with-glibc2.35', 'Environment': 'Linux', 'Python': '3.10.12', 'Install': 'pip', 'RAM': '31.24 GB', 'Disk': '39.1/250.9 GB', 'CPU': '12th Gen Intel Core(TM) i5-12600', 'CPU count': 12, 'GPU': 'NVIDIA GeForce RTX 3060, 12288MiB', 'GPU count': 1, 'CUDA': '12.4', 'Package Info': {'numpy': '✅ 1.26.4<2.0.0; sys_platform == "darwin"', 'matplotlib': '✅ 3.8.3>=3.3.0', 'opencv-python': '✅ 4.10.0.84>=4.6.0', 'pillow': '✅ 10.2.0>=7.1.2', 'pyyaml': '✅ 5.4.1>=5.3.1', 'requests': '✅ 2.32.3>=2.23.0', 'scipy': '✅ 1.14.1>=1.4.1', 'torch': '✅ 2.5.1!=2.4.0,>=1.8.0; sys_platform == "win32"', 'torchvision': '✅ 0.20.1>=0.9.0', 'tqdm': '✅ 4.66.2>=4.64.0', 'psutil': '✅ 6.0.0', 'py-cpuinfo': '✅ 9.0.0', 'pandas': '✅ 2.2.3>=1.1.4', 'seaborn': '✅ 0.13.2>=0.11.0', 'ultralytics-thop': '✅ 2.0.13>=2.0.0'}}

### Minimal Reproducible Example

```python
import torch 
from .metrics import bbox_iou

box1 = torch.tensor([[0.5, 0.5, 1.0, 1.0], [0.3, 0.3, 0.8, 0.8]])  # Shape (2, 4)
box2 = torch.tensor([[0.4, 0.4, 0.9, 0.9]])  # Shape (1, 4)
iou = bbox_iou(box1, box2, xywh=False)
print(iou)  # Outputs IoU values for both bounding boxes in `box1`
```

### Additional

no

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @visionNoob, thank you for your interest in Ultralytics 🚀! We appreciate you taking the time to report this issue and dig into the specifics of the `bbox_iou` function. Your detailed explanation and example are super helpful! 🔎

We recommend a visit to the [Docs](https://docs.ultralytics.com), where you might find relevant information or clarifications. For future reference, if this issue is related to a 🐛 bug, please ensure you provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/), which you've already done an excellent job of including here—thank you!

If this is a ❓ documentation-related query and you’re willing to help with improvements, your offer to submit a PR is very welcome! Please ensure you follow our [Contribution Guide](https://github.com/ultralytics/ultralytics/blob/main/CONTRIBUTING.md) for submitting changes smoothly.

### Upgrade 

Just a quick note to ensure you're running the latest version of Ultralytics and its dependencies for consistency across implementations and verification. You can upgrade using:
```bash
pip install -U ultralytics
```

### Community

You can also explore our Ultralytics community to share knowledge or seek additional support:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time support and chats 
- [Discourse](https://community.ultralytics.com) for in-depth discussions 
- [Subreddit](https://reddit.com/r/Ultralytics) to interact with the larger Ultralytics community

### Environments 

You may want to test your work in our verified environments to ensure consistency:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
  
### Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is 💚 green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing. CI tests verify the operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu.

This is an automated response—an Ultralytics engineer will be with you shortly to assist further 😊. Thank you for contributing!

**Comment by Y-T-G** [comment]
You can open a PR to update the docstring

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #18579](https://github.com/ultralytics/ultralytics/pull/18579) 🎉. Huge thanks to the PR author @visionNoob and contributors @Y-T-G and @glenn-jocher for their efforts in implementing this solution.

### ✨ Key Changes
The `bbox_iou` function has been enhanced to support a broader range of input shapes, offering more flexibility for bounding box IoU computations. The updated function now accepts tensors with shapes such as `(4,)`, `(N, 4)`, `(B, N, 4)`, and `(B, N, 1, 4)` as long as the last dimension is `4`. The logic has been updated to automatically adapt to these formats, making it easier to work with diverse datasets and reducing pre-processing needs. Additionally, the changes improve functionality for both single and batched IoU evaluations, streamlining workflows and saving time.

### 🔍 Testing the Fix
To test whether the changes resolve this issue, you can try out the latest updates directly by running:
```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Alternatively, you can wait for the next official release on PyPI.  

Please let us know if this fix resolves your issue! Your feedback is invaluable as it helps us ensure the changes are effective and supports further improvements.  

Thank you for reporting this issue and helping make the project better 🙏. Feel free to share any additional feedback if the problem persists!
